File: src/theme.ts
// Palette derived from research on Spike Jonze's "Her" (production design by
// K.K. Barrett, cinematography by Hoyte van Hoytema) and adapted for a 2-player
// binary-star game. Warm tones throughout; no blues, no greens, no neon.
//
// Source: docs/research/her-aesthetic.md (research report, 2026-06-01).

export const palette = {
  voidDeep: '#1A0F14', // background — deep wine-black, never pure black
  player1: '#E8956F', // warm peach-coral
  player2: '#D97D3D', // burnt apricot-rust
  rose: '#F4A58D', // UI text, soft halos
  cream: '#FFC89B', // highlights, star cores
  terracotta: '#A3685C', // court lines, secondary UI, orbital trails
  wine: '#6F1D1B', // accent shadows, "unbound" warning text
  // Bright warning red for the one peril affordance — the "Add 3rd Body"
  // button + the intruding third star in the post-win three-body unravel. A
  // deliberate hot break from the warm palette: it MUST read as danger.
  danger: '#E5392F',
  // The pale blue dot — a world, a deliberate cool note among the warm suns.
  // Muted so it reads as a small distant world, not neon.
  world: '#7FB0C9',
} as const;

export const fonts = {
  // Humanist serif for wordmark, headings, victory cards
  serif: '"Cardo", Georgia, "Times New Roman", serif',
  // Humanist sans for UI labels, tooltips, px/s readout
  sans: '"Inter", system-ui, -apple-system, "Segoe UI", sans-serif',
} as const;

// ─── Legibility floor for contain-fit canvas text ──────────────────────────
//
// The game draws in a fixed design space (1280×800 / 800×1280) that the
// Renderer maps into the viewport with a uniform contain-fit. A phone held in
// portrait renders the 800×1280 space at ~0.49×, so a 12px design font lands
// at ~5.9 CSS px on screen — far below the ~11px readability floor. (Found on
// the iOS port first; the web has identical fit math.)
//
// `cpx` pulls any design size that would render below the floor up toward it,
// compressing the sub-floor range (slope) so the size hierarchy is preserved,
// never inverted. Text already at/above the floor — all display type, and any
// near-1× desktop fit — is returned unchanged. The Renderer calls
// `setViewScale(fit.scale)` once per frame before any drawing or measuring.
let viewScale = 1;
const FLOOR_PX = 11; // on-screen CSS px (Apple HIG-ish minimum)
const SLOPE = 0.2; // fraction of the sub-floor deficit that survives

export function setViewScale(scale: number): void {
  viewScale = scale > 0 ? scale : 1;
}

// Design-space px to draw at so the glyph clears the on-screen floor. Identity
// at/above the floor — continuous at the boundary, monotonic in designSize.
export function cpx(designSize: number): number {
  const onScreen = designSize * viewScale;
  if (onScreen >= FLOOR_PX) return designSize;
  return (FLOOR_PX - (FLOOR_PX - onScreen) * SLOPE) / viewScale;
}

// Line advance for compensated text — call sites hardcode design-px line
// heights tuned for the UNcompensated sizes; this keeps stacked/wrapped text
// from overlapping once it grows. 1.35 matches the iOS port.
export function lineHeightFor(designSize: number): number {
  return cpx(designSize) * 1.35;
}

export type Palette = typeof palette;
export type PaletteKey = keyof Palette;

// Convert a hex color to an rgba string with the given alpha. Used everywhere
// we need a translucent version of a palette color (glows, trails, overlays).
export function rgba(hex: string, alpha: number): string {
  const h = hex.replace('#', '');
  const r = parseInt(h.slice(0, 2), 16);
  const g = parseInt(h.slice(2, 4), 16);
  const b = parseInt(h.slice(4, 6), 16);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}

// Linear-blend two hex colors in sRGB. t=0 returns a, t=1 returns b.
// Used by the Doppler tint code so a star's primary color shifts toward
// cream (approaching the observer) or wine (receding) without leaving
// the palette.
export function blendHex(a: string, b: string, t: number): string {
  const ah = a.replace('#', '');
  const bh = b.replace('#', '');
  const ar = parseInt(ah.slice(0, 2), 16);
  const ag = parseInt(ah.slice(2, 4), 16);
  const ab = parseInt(ah.slice(4, 6), 16);
  const br = parseInt(bh.slice(0, 2), 16);
  const bg = parseInt(bh.slice(2, 4), 16);
  const bb = parseInt(bh.slice(4, 6), 16);
  const clamp01 = (x: number): number => (x < 0 ? 0 : x > 1 ? 1 : x);
  const k = clamp01(t);
  const r = Math.round(ar + (br - ar) * k);
  const g = Math.round(ag + (bg - ag) * k);
  const bl = Math.round(ab + (bb - ab) * k);
  const hex = (n: number): string => n.toString(16).padStart(2, '0');
  return `#${hex(r)}${hex(g)}${hex(bl)}`.toUpperCase();
}

File: ios/App/Theme.swift
import SwiftUI

// The "Her" warm palette — verbatim hexes from src/theme.ts. Warm tones
// throughout; no blues, no greens, no neon. Every color in the app comes
// from here.
enum Palette {
  static let voidDeep = Color(hex: 0x1A0F14)
  static let player1 = Color(hex: 0xE8956F)
  static let player2 = Color(hex: 0xD97D3D)
  static let rose = Color(hex: 0xF4A58D)
  static let cream = Color(hex: 0xFFC89B)
  static let terracotta = Color(hex: 0xA3685C)
  static let wine = Color(hex: 0x6F1D1B)
}

extension Color {
  init(hex: UInt32) {
    self.init(
      .sRGB,
      red: Double((hex >> 16) & 0xFF) / 255,
      green: Double((hex >> 8) & 0xFF) / 255,
      blue: Double(hex & 0xFF) / 255,
      opacity: 1
    )
  }
}

/// Linear sRGB blend between two palette colors — the Doppler tint mechanism
/// (port of theme.ts blendHex). t = 0 → a, t = 1 → b.
func blendColor(_ a: Color, _ b: Color, _ t: Double) -> Color {
  let k = min(max(t, 0), 1)
  let ca = UIColor(a), cb = UIColor(b)
  var ar: CGFloat = 0, ag: CGFloat = 0, ab2: CGFloat = 0, aa: CGFloat = 0
  var br: CGFloat = 0, bg: CGFloat = 0, bb: CGFloat = 0, ba: CGFloat = 0
  ca.getRed(&ar, green: &ag, blue: &ab2, alpha: &aa)
  cb.getRed(&br, green: &bg, blue: &bb, alpha: &ba)
  return Color(
    .sRGB,
    red: ar + (br - ar) * k,
    green: ag + (bg - ag) * k,
    blue: ab2 + (bb - ab2) * k,
    opacity: 1
  )
}

// Legibility compensation for contain-fit text (QA finding, 2026-06-10,
// found by JP on an iPhone 14): the 800×1280 portrait design space lands on
// iPhone screens at ~0.49× scale, so every design-px font renders at half
// size — 12px help text became ~5.9pt against Apple's ~11pt legibility
// floor. Fonts are drawn inside the fit-scaled canvas context, so the
// compensation happens here in design units: text that would land below
// `floorPt` on screen is pulled up toward it, compressing (slope) but never
// reversing the size hierarchy; text already at/above the floor — display
// type, and ALL text on large-scale fits like iPad (~0.92×) — is untouched.
// Continuous at the floor boundary by construction.
enum Typography {
  /// Live contain-fit scale, written by ContentView.render each frame.
  /// 0 (pre-first-frame) means "no compensation".
  static var fitScale: CGFloat = 0
  static let floorPt: CGFloat = 11 // on-screen points
  static let slope: CGFloat = 0.2  // how much sub-floor deficit survives

  static func compensate(_ designSize: CGFloat) -> CGFloat {
    guard fitScale > 0 else { return designSize }
    let pt = designSize * fitScale
    guard pt < floorPt else { return designSize }
    return (floorPt - (floorPt - pt) * slope) / fitScale
  }

  /// Line advance for compensated text — call sites hardcode design-px line
  /// heights tuned for UNcompensated sizes; this keeps wrapped/stacked text
  /// from overlapping once it grows. 1.35 leading: tight enough that the
  /// 3-line setup help clears the court's top edge and the Carse footer
  /// stays inside the win card (verified by /ios-qa screenshots).
  static func lineHeight(for designSize: CGFloat) -> CGFloat {
    compensate(designSize) * 1.35
  }
}

// Cardo (serif: wordmark, cards) + Inter (sans: UI labels) — both SIL OFL,
// bundled as app resources. Fall back to the system designs so a missing
// font file degrades gracefully instead of crashing. Sizes pass through
// Typography.compensate (legibility floor) — call sites stay in design px.
enum Fonts {
  static func serif(_ size: CGFloat, italic: Bool = false) -> Font {
    let s = Typography.compensate(size)
    let name = italic ? "Cardo-Italic" : "Cardo-Regular"
    if UIFont(name: name, size: s) != nil {
      return .custom(name, size: s)
    }
    let f = Font.system(size: s, design: .serif)
    return italic ? f.italic() : f
  }

  static func sans(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
    let s = Typography.compensate(size)
    if UIFont(name: "Inter-Regular", size: s) != nil {
      return .custom("Inter-Regular", size: s).weight(weight)
    }
    return .system(size: s, weight: weight)
  }
}
